pkg/util/platformworkloadidentity/platformworkloadidentityrolesbyversion.go

```diff
@@ -0,0 +1,81 @@
+package platformworkloadidentity
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+// PlatformWorkloadIdentityRolesByVersion is the interface that validates and obtains the version from an PlatformWorkloadIdentityRoleSetDocument.
+type PlatformWorkloadIdentityRolesByVersion interface {
+	GetPlatformWorkloadIdentityRolesByRoleName() map[string]api.PlatformWorkloadIdentityRole
+}
+
+// platformWorkloadIdentityRolesByVersionService is the default implementation of the PlatformWorkloadIdentityRolesByVersion interface.
+type platformWorkloadIdentityRolesByVersionService struct {
+	platformWorkloadIdentityRoles []api.PlatformWorkloadIdentityRole
+}
+
+func NewPlatformWorkloadIdentityRolesByVersion(ctx context.Context, oc *api.OpenShiftCluster, dbPlatformWorkloadIdentityRoleSets database.PlatformWorkloadIdentityRoleSets) (PlatformWorkloadIdentityRolesByVersion, error) {
+	if !oc.UsesWorkloadIdentity() {
+		return nil, nil
+	}
+
+	currentInstallVersion := oc.Properties.ClusterProfile.Version
+	currentOpenShiftVersion, err := version.ParseVersion(currentInstallVersion)
+	if err != nil {
+		return nil, err
+	}
+	currentMinorVersion := currentOpenShiftVersion.MinorVersion()
+	requiredMinorVersions := map[string]bool{currentMinorVersion: false}
+	platformWorkloadIdentityRoles := []api.PlatformWorkloadIdentityRole{}
+
+	docs, err := dbPlatformWorkloadIdentityRoleSets.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo != nil {
+		upgradeableVersion, err := version.ParseVersion(string(*oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo))
+		if err != nil {
+			return nil, err
+		}
+		upgradeableMinorVersion := upgradeableVersion.MinorVersion()
+		if currentMinorVersion != upgradeableMinorVersion && !currentOpenShiftVersion.Lt(upgradeableVersion) {
+			requiredMinorVersions[upgradeableMinorVersion] = false
+		}
+	}
+
+	for _, doc := range docs.PlatformWorkloadIdentityRoleSetDocuments {
+		for version := range requiredMinorVersions {
+			if version == doc.PlatformWorkloadIdentityRoleSet.Properties.OpenShiftVersion {
+				platformWorkloadIdentityRoles = append(platformWorkloadIdentityRoles, doc.PlatformWorkloadIdentityRoleSet.Properties.PlatformWorkloadIdentityRoles...)
+				requiredMinorVersions[version] = true
+			}
+		}
+	}
+
+	for version, exists := range requiredMinorVersions {
+		if !exists {
+			return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "No PlatformWorkloadIdentityRoleSet found for the requested or upgradeable OpenShift minor version '%s'. Please retry with different OpenShift version, and if the issue persists, raise an Azure support ticket", version)
```

Comment:
I think we'll need a more clarifying error message here, as the message here as-is might be confusing to users. Is this returned when an invoking user does not provide a required identity for a given OCP x.y.z version but passes that version in as UpgradeableTo?

Reply:
This logic works for getting the PlatformWorkloadIdentityRoleSet for the OCP minor version stated in clusterprofile or UpgradeableTo. It has nothing to do if required identities are provided or not.